server/src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java

```diff
@@ -123,8 +124,17 @@ private StoreFilesMetaData listStoreMetaData(ShardId shardId) throws IOException
             if (indexService != null) {
                 IndexShard indexShard = indexService.getShardOrNull(shardId.id());
                 if (indexShard != null) {
-                    exists = true;
-                    return new StoreFilesMetaData(shardId, indexShard.snapshotStoreMetadata());
+                    try {
+                        final StoreFilesMetaData storeFilesMetaData = new StoreFilesMetaData(shardId, indexShard.snapshotStoreMetadata());
+                        exists = true;
+                        return storeFilesMetaData;
+                    } catch (org.apache.lucene.index.IndexNotFoundException e) {
+                        logger.trace(new ParameterizedMessage("[{}] node is missing index, responding with empty", shardId), e);
+                        return new StoreFilesMetaData(shardId, Store.MetadataSnapshot.EMPTY);
```

Comment:
Should we only ignore the exception if shard is recovering?

Reply:
Would be nice to only have this try-catch in the no-engine case, since it should not happen if there is an engine. The same pattern is also done in 2 other places already, so could be nice to add a single method to handle this in the same way whenever needed.
I see no problems in doing this always in this specific case, so please regard this optional.

Reply:
I will leave this to a follow-up to explore. I'm generally not happy with how we access the store on a shard that is recovering. Also accessing last commit instead of safe commit feels wrong as well for some of the callers.